### PR TITLE
multiple_crate_versions: skip dev and build deps

### LINKED ISSUE #5041: multiple_crate_versions lint detects dev dependencies also
The stated reason for the multiple_crate_versions check is 

> This bloats the size of targets, and can lead to confusing error messages when structs or traits are used interchangeably between different versions of a crate.

However this error triggers when multiple crates declare a dev dependency on the same target also. For example:
```
warning: multiple versions for dependency `syn`: 0.15.44, 1.0.11
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions
```
However because this is a dev dependency it should not cause any of the issues mentioned above. It would be good to exclude dev and test dependencies from this check.

### PR DESCRIPTION
**Pull request by ebroto**
changelog: multiple_crate_versions: skip dev and build deps

Closes #5041

### CODE REVIEW

File: clippy_lints/src/multiple_crate_versions.rs [hunk 1]
```diff
@@ -34,37 +37,65 @@ declare_clippy_lint! {
...
 
+        let local_name = cx.tcx.crate_name(LOCAL_CRATE).as_str();
         let mut packages = metadata.packages;
         packages.sort_by(|a, b| a.name.cmp(&b.name));
 
-        for (name, group) in &packages.into_iter().group_by(|p| p.name.clone()) {
-            let group: Vec<cargo_metadata::Package> = group.collect();
+        if_chain! {
+            if let Some(resolve) = &metadata.resolve;
+            if let Some(local_id) = packages.iter().find(|p| p.name == *local_name).map(|p| &p.id);
```
[the remark this answers is not in the document]
Comment (ebroto) [reply clippy_lints/src/multiple_crate_versions.rs]: Nothing important, just thought that the `if_chain` was more readable if each entry fitted in one line. I guess it doesn't matter, I've changed it :)

**Comment by phansch** [comment]
Ah that should have been

@bors r+

**Pull request merged**